File: src/Nexo.Tests.Infrastructure/scripts/copy-assemblies.py
#!/usr/bin/env python3
"""Parse deps.json and copy all required assemblies to output directory."""
import json
import os
import sys
import shutil

def main():
    if len(sys.argv) < 3:
        print("Usage: copy-assemblies.py <deps.json> <output-dir> [nuget-packages-root]")
        sys.exit(1)
    
    deps_json_path = sys.argv[1]
    output_dir = sys.argv[2]
    nuget_root = sys.argv[3] if len(sys.argv) > 3 else os.path.expanduser("~/.nuget/packages")
    
    if not os.path.exists(deps_json_path):
        print(f"deps.json not found: {deps_json_path}")
        sys.exit(1)
    
    with open(deps_json_path, 'r') as f:
        deps = json.load(f)
    
    targets = deps.get('targets', {})
    runtime = targets.get('.NETCoreApp,Version=v8.0', {})
    
    copied = 0
    failed = 0
    
    for lib_name, lib_data in runtime.items():
        if '/' not in lib_name:
            continue
        
        runtime_paths = lib_data.get('runtime', {})
        for path, info in runtime_paths.items():
            if not path.endswith('.dll'):
                continue
            
            # Extract package name and version
            pkg_name, pkg_version = lib_name.split('/')
            pkg_name_lower = pkg_name.lower()
            
            # Build source path
            lib_path = path.replace('/', os.sep)
            source_path = os.path.join(nuget_root, pkg_name_lower, pkg_version, lib_path)
            
            # Build destination path
            dll_name = os.path.basename(path)
            dest_path = os.path.join(output_dir, dll_name)
            
            # Copy if source exists
            if os.path.exists(source_path):
                try:
                    shutil.copy2(source_path, dest_path)
                    copied += 1
                except Exception as e:
                    print(f"Failed to copy {source_path}: {e}", file=sys.stderr)
                    failed += 1
            else:
                # Don't fail on missing assemblies - some might be in different locations
                pass
    
    print(f"Copied {copied} assemblies, {failed} failures")
    return 0 if failed == 0 else 1

if __name__ == '__main__':
    sys.exit(main())

File: src/Nexo.Tests.Infrastructure/scripts/copy-assemblies.cs
// Parse deps.json and copy all required assemblies to output directory.
// This is a C# program that can be compiled and run with: dotnet run --project <path> -- <args>
// Or compiled to a single file and executed.

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using System.Linq;

if (args.Length < 2)
{
    Console.Error.WriteLine("Usage: copy-assemblies <deps.json> <output-dir> [nuget-packages-root]");
    Environment.Exit(1);
}

var depsJsonPath = args[0];
var outputDir = args[1];
// Windows callers can pass an escaped quote suffix when an output path ends in '\' and
// the argument itself is quoted by MSBuild Exec. Trim any accidental wrapping quotes and
// normalize trailing separators so directory creation always receives a valid path.
outputDir = outputDir.Trim().Trim('"');
outputDir = outputDir.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
if (string.IsNullOrWhiteSpace(outputDir))
{
    Console.Error.WriteLine("Output directory argument is empty after normalization.");
    Environment.Exit(1);
}
var nugetRoot = args.Length > 2 ? args[2] : Path.Combine(
    Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
    ".nuget", "packages");

if (!File.Exists(depsJsonPath))
{
    Console.Error.WriteLine($"deps.json not found: {depsJsonPath}");
    Environment.Exit(1);
}

// Ensure output directory exists
Directory.CreateDirectory(outputDir);

// Parse deps.json
string jsonContent = File.ReadAllText(depsJsonPath);
using JsonDocument doc = JsonDocument.Parse(jsonContent);
var root = doc.RootElement;

if (!root.TryGetProperty("targets", out var targetsElement))
{
    Console.Error.WriteLine("deps.json does not contain 'targets' property");
    Environment.Exit(1);
}

// Find the runtime target (use runtimeTarget from root, or first NETCoreApp target)
string? targetName = null;
if (root.TryGetProperty("runtimeTarget", out var rtElement) && rtElement.TryGetProperty("name", out var rtName))
    targetName = rtName.GetString();

JsonElement runtime;
if (!string.IsNullOrEmpty(targetName) && targetsElement.TryGetProperty(targetName, out var rtValue))
    runtime = rtValue;
else
{
    var first = targetsElement.EnumerateObject().FirstOrDefault(p => p.Name.Contains("NETCoreApp", StringComparison.OrdinalIgnoreCase));
    if (first.Value.ValueKind == JsonValueKind.Undefined)
    {
        Console.Error.WriteLine("Could not find runtime target in deps.json");
        Environment.Exit(1);
    }
    runtime = first.Value;
}
int copied = 0;
int failed = 0;

foreach (var libProperty in runtime.EnumerateObject())
{
    var libName = libProperty.Name;
    
    // Skip if not a package reference (package references have '/' separator)
    if (!libName.Contains('/'))
        continue;
    
    // Extract package name and version
    var parts = libName.Split('/', 2);
    if (parts.Length != 2)
        continue;
    
    var pkgName = parts[0];
    var pkgVersion = parts[1];
    var pkgNameLower = pkgName.ToLowerInvariant(); // NuGet package folders are lowercase
    
    // Collect runtime assembly paths from both "runtime" and "runtimeTargets".
    // On Windows RID-specific graphs, dependencies are often emitted in runtimeTargets.
    // IMPORTANT: Some packages (e.g. System.Text.Encodings.Web) list both lib/netX and
    // runtimes/browser/... assets. Copying the browser build last overwrites the desktop DLL
    // and breaks System.Text.Json (PlatformNotSupportedException in JsonSerializer .cctor).
    var currentRid = System.Runtime.InteropServices.RuntimeInformation.RuntimeIdentifier;
    var candidatesByDll = new Dictionary<string, List<(string RelPath, int Priority)>>(StringComparer.OrdinalIgnoreCase);

    void AddCandidate(string relPath, int priority)
    {
        if (!relPath.EndsWith(".dll", StringComparison.OrdinalIgnoreCase))
            return;
        // Never ship browser/wasm builds into desktop test output.
        if (relPath.Contains("/runtimes/browser/", StringComparison.OrdinalIgnoreCase) ||
            relPath.Contains("\\runtimes\\browser\\", StringComparison.OrdinalIgnoreCase))
            return;

        var dllName = Path.GetFileName(relPath);
        if (!candidatesByDll.TryGetValue(dllName, out var list))
        {
            list = new List<(string, int)>();
            candidatesByDll[dllName] = list;
        }
        list.Add((relPath, priority));
    }

    if (libProperty.Value.TryGetProperty("runtime", out var runtimePaths))
    {
        foreach (var runtimePathProperty in runtimePaths.EnumerateObject())
            AddCandidate(runtimePathProperty.Name, priority: 0); // lib/ assets win over RID-specific
    }

    if (libProperty.Value.TryGetProperty("runtimeTargets", out var runtimeTargets))
    {
        foreach (var runtimeTargetProperty in runtimeTargets.EnumerateObject())
        {
            if (!runtimeTargetProperty.Value.TryGetProperty("assetType", out var assetType) ||
                !string.Equals(assetType.GetString(), "runtime", StringComparison.OrdinalIgnoreCase))
                continue;

            var relPath = runtimeTargetProperty.Name;
            if (runtimeTargetProperty.Value.TryGetProperty("rid", out var ridEl))
            {
                var rid = ridEl.GetString() ?? "";
                if (string.Equals(rid, "browser", StringComparison.OrdinalIgnoreCase))
                    continue;
                if (!string.Equals(rid, currentRid, StringComparison.OrdinalIgnoreCase))
                    continue;
            }

            AddCandidate(relPath, priority: 1);
        }
    }

    foreach (var kv in candidatesByDll)
    {
        var best = kv.Value.OrderBy(t => t.Priority).ThenBy(t => t.RelPath, StringComparer.OrdinalIgnoreCase).First();
        var relPath = best.RelPath;
        var libPath = relPath.Replace('/', Path.DirectorySeparatorChar);
        var sourcePath = Path.Combine(nugetRoot, pkgNameLower, pkgVersion, libPath);
        var destPath = Path.Combine(outputDir, kv.Key);

        if (!File.Exists(sourcePath))
            continue;

        try
        {
            File.Copy(sourcePath, destPath, overwrite: true);
            copied++;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Failed to copy {sourcePath}: {ex.Message}");
            failed++;
        }
    }
}

Console.WriteLine($"Copied {copied} assemblies, {failed} failures");
Environment.Exit(failed == 0 ? 0 : 1);
